api/v1alpha1/module_types.go

```diff
@@ -92,42 +93,112 @@ type KernelMapping struct {
-	// Name is ignored and is set automatically by the OOT Operator.
-	DevicePlugin *v1.Container `json:"devicePlugin"`
+	Build *Build `json:"build,omitempty"`
 
-	// DriverContainer allows overriding some properties of the container that deploys the driver on the node.
-	// Name and image are ignored and are set automatically by the OOT Operator.
-	DriverContainer v1.Container `json:"driverContainer"`
+	// ContainerImage is a top-level field
+	// +optional
+	ContainerImage string `json:"containerImage,omitempty"`
+
+	// Image pull policy.
+	// One of Always, Never, IfNotPresent.
+	// Defaults to Always if :latest tag is specified, or IfNotPresent otherwise.
+	// Cannot be updated.
+	// More info: https://kubernetes.io/docs/concepts/containers/images#updating-images
+	// +optional
+	ImagePullPolicy v1.PullPolicy `json:"imagePullPolicy,omitempty" protobuf:"bytes,14,opt,name=imagePullPolicy,casttype=PullPolicy"`
 
 	// KernelMappings is a list of kernel mappings.
 	// When a node's labels match Selector, then the OOT Operator will look for the first mapping that matches its
 	// kernel version, and use the corresponding container image to run the DriverContainer.
+	// +kubebuilder:validation:MinItems=1
 	KernelMappings []KernelMapping `json:"kernelMappings"`
 
-	// Selector describes on which nodes the Module should be loaded.
-	Selector map[string]string `json:"selector"`
+	// Modprobe is a set of properties to customize which module modprobe loads and with which properties.
+	Modprobe ModprobeSpec `json:"modprobe"`
+}
+
+type DriverContainerSpec struct {
+	// Container holds the properties for the driver container that runs modprobe.
+	Container DriverContainerContainerSpec `json:"container"`
+
+	// +optional
+	ImageRepoSecret *v1.LocalObjectReference `json:"imageRepoSecret,omitempty"`
+
+	// +optional
+	// ServiceAccountName is the name of the ServiceAccount to use to run this pod.
+	// More info: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+}
+
+type DevicePluginContainerSpec struct {
+	// Image is the name of the container image that the device plugin container will run.
+	Image string `json:"image"`
+
+	// VolumeMounts is a list of volume mounts that are appended to the default ones.
+	// +optional
+	VolumeMounts []v1.VolumeMount `json:"volumeMounts,omitempty"`
+}
+
+type DevicePluginSpec struct {
+	Container DevicePluginContainerSpec `json:"container"`
+
+	ImagePullSecrets []v1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
 
 	// +optional
 	// ServiceAccountName is the name of the ServiceAccount to use to run this pod.
 	// More info: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/
 	ServiceAccountName string `json:"serviceAccountName,omitempty"`
 
+	Volumes []v1.Volume `json:"volumes,omitempty"`
```

Comment:
Although in k8s there is a reason for having `VolumeMounts` and `Volume` in 2 different objects (the reason is that a `Volume` is added at the pod level but for each container it need to be mapped to a `VolumeMount` location), I don't see the reason for it here.

As I wrote for `DriverContainerSpec`, here as well I don't see a logical separation between `DevicePluginSpec` and `DevicePluginContainerSpec`.

What is the reason for those separations?

Reply:
Same as above. `DevicePluginSpec` contains all properties related to the pod, while `DevicePluginContainerSpec` contains all properties related to the first container in that pod.

Reply:
Will you suggest having them all under a single object `DevicePluginSpec`. I think it will be less confusing. It feels like we are enforcing our CRD to get the pod/container structure but those are different objects eventually.

Reply:
> It feels like we are enforcing our CRD to get the pod/container structure

Yes. If for any reason we need to add properties for a second container in the pod, we just need to add another sub-object - there is a clear separation between what belongs to the pod and what belongs to the container.

> but those are different objects eventually.

Not sure I understand your point here.